widget/editor.go

```diff
@@ -99,6 +102,27 @@ type maskReader struct {
 	overflow []byte
 }
 
+// combinedPos is a point in the editor, three different ways.
+type combinedPos struct {
+	// editorBuffer offset. The other three fields are based off of this one.
+	ofs int
+
+	// lineCol.Y = line (offset into Editor.lines), and X = col (offset into
+	// Editor.lines[Y])
+	lineCol screenPos
+
+	// Pixel coordinates
+	x fixed.Int26_6
+	y int
+}
+
+type SelectionAction int
+
+const (
+	SelectionClear SelectionAction = iota
+	SelectionExtend
+)
```

Comment:
```go
> // Move is the relative variant of Move. It moves the caret end bytes from its current position, and
> // the selection start bytes from its current position.
> func (e *Editor) Move(start, end int)
```

I assume you mean `Move is the relative variant of MoveAbs`?

TBH I haven't looked closely at CaretCoord or CaretPos, since they seemed mostly unrelated to what I was doing.

It does seem like if we have Caret() which gets the caret and selection positions, most other APIs would have SetCaret (i.e. Thing() to get "thing(s)" and SetThing(...) to set it/them).  So what do you think of that?  Caret() to get and SetCaret() (replacing Move) for relative movement, and SetCaretAbs() (replacing MoveAbs) for absolute movement?

I have some concerns about Move()/SetCaret(), but I think we need to get some feedback on how people want to use it, so I'm willing to do as you suggest right now.

Reply:
> ```go
> > // Move is the relative variant of Move. It moves the caret end bytes from its current position, and
> > // the selection start bytes from its current position.
> > func (e *Editor) Move(start, end int)
> ```
> 
> I assume you mean `Move is the relative variant of MoveAbs`?
> 

Yes. Sorry about that.

> TBH I haven't looked closely at CaretCoord or CaretPos, since they seemed mostly unrelated to what I was doing.
> 

Me neither, but I noticed that you added more API and had to consider whether it fits.

> It does seem like if we have Caret() which gets the caret and selection positions, most other APIs would have SetCaret (i.e. Thing() to get "thing(s)" and SetThing(...) to set it/them). So what do you think of that? Caret() to get and SetCaret() (replacing Move) for relative movement, and SetCaretAbs() (replacing MoveAbs) for absolute movement?
> 

I wouldn't go so far in normalizing method names. For example SetCaret sounds like your SetCaretAbs to me and is a nicer name. What about "SetCaret" (absolute) and "MoveCaret" (relative)?

> I have some concerns about Move()/SetCaret(), but I think we need to get some feedback on how people want to use it, so I'm willing to do as you suggest right now.

Selection is subtle enough, which is why I suggested keeping Selection unexported and leaving the rest of the API alone, postponing the xported API to future changes. However, I can understand why you want the API in this change: the exported API can often guide simplifications in the internal implementation.

Reply:
> What about "SetCaret" (absolute) and "MoveCaret" (relative)?

SGTM.

> ... exported api discussion ...

We can't keep the entirety of the selection part of the api private because we use Move internally.

Reply:
The (new) `MoveCaret(startDelta, endDelta int)` (which replaces `Move(distance int, selAct SelectionAction)`) has me, I don't know, looking askance at it.  In particular, I don't like that the caret is the end of the selection and not the start.  Would it be awful if I swapped them?  That is, made the caret always the start of the selection and change `selStart` to `selEnd`?  WDYT?

Reply:
Yes, please! I also found the caret being the end of the selection.

Perhaps we could go all the way and consider a caret "fat", having a start and end coordinate? That would fit better with the MoveCaret method, and further enforces the approach of always having a selection even though it is usually zero runes.

Reply:
> Perhaps we could go all the way and consider a caret "fat", having a start and end coordinate?

I feel like, operationally speaking, that's what we have now.  I guess we could make it more obvious / formal / explicit and change caret.pos to caret.start, and change/move selStart to caret.end.  That makes sense.

Also, in case it's not obvious: the big reason I've avoided only using an offset, and actually storing the line & column coordinates, and the x,y pixel coordinates, is for efficiency.  I figure that if I only stored offsets, then for every single layout we'd have to recalculate their coordinates within the widget, and that seemed like it could get prohibitive, especially with a lot of text, during a mouse drag or a resize, where we're doing layouts hand-over-fist.

Reply:
> > In particular, I don't like that the caret is the end of the selection and not the start. Would it be awful if I swapped them?
>
> Yes, please!

Done.